Multi-handle slider implementation

cuberoot opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

Refactor sp-slider to allow for explicit declaration of handles, including multiple handles as per #1385.

## Related Issue

fixes #1385

## Motivation and Context

There are a number of use cases for multi-handle sliders in creative apps, including range-slider contribution to Spectrum.

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

cuberoot commented:
@Westbrook Here's my work in progress so far. It is getting close. I haven't implemented the slider handle constraints yet and I need to work on the tests. It would be great to get early feedback.

Westbrook commented on packages/slider/src/Slider.ts:
```diff
@@ -16,44 +16,69 @@ import {
 
-        if (newValue === oldValue) {
-            return;
+    @property({ reflect: true, converter: ValueConverter })
```
I've had trouble with `@property` working as expected (usually as documented, but you know expectations...) with getter/setter usage. If this works, great, but just something to keep an eye on it.

cuberoot replied:
Good to know. I think I got the idea from the original slider code. I will watch it like a hawk 😄

Westbrook replied:
It was mainly that either reflection or render didn't happen without actually telling it to. Might be good to focus a test on those two side effects here.

cuberoot replied:
I changed to make `value` a simple number attribute

Westbrook commented on packages/slider/src/Slider.ts:
```diff
@@ -16,44 +16,69 @@ import {
 
-        this._value = newValue;
-        this.requestUpdate('value', oldValue);
+    public set value(value: number | HandleValues) {
```
I would wonder if `this.handleController.hasExplicitHandles && this.handleController.count` could handle `isNumber` and if `!this.handleController.hasExplicitHandles` could handle `HandleValues.length === 1`. Some logic is too much...but could be interesting.

cuberoot replied:
I changed the value attribute to be simpler

Westbrook commented on packages/slider/src/Slider.ts:
```diff
@@ -78,18 +103,29 @@ export class Slider extends Focusable {
+        const valueArray = Array.from(values.values());
+        if (valueArray.length === 2)
+            return `${valueArray[0]} - ${valueArray[1]}`;
+        return valueArray.join(', ');
```
This seems like it'll need some thinking...you're likely already doing that with usage of this commented out in the handle controller.

Westbrook replied:
This doesn't exactly feel related to this line anymore, any thoughts on what I was going on about here?

cuberoot replied:
🤔 Not sure what this was about

Westbrook commented on packages/slider/src/Slider.ts:
```diff
@@ -78,18 +103,29 @@ export class Slider extends Focusable {
+        return valueArray.join(', ');
+    };
 
     @property({ attribute: false })
```
Not your code, but not sure why this is `@property`...

cuberoot replied:
Will try removing the `@property`

Westbrook commented on packages/slider/src/Slider.ts:
```diff
@@ -143,51 +165,39 @@ export class Slider extends Focusable {
+    public connectedCallback(): void {
+        super.connectedCallback();
+        this.handleController.hostConnected();
+        this.addEventListener(
```
I feel like this is an implementation for more flexible reuse, which is cool, but we might want to just allow the controller too access `this.host.requestUpdate()` to simplify this for our use case?

cuberoot replied:
Are you talking about the event listener? It is listening to the slider handle, which the user owns in case they set the value programatically.

Westbrook replied:
Ah, ok,  that makes sense.

Westbrook commented on packages/slider/src/Slider.ts:
```diff
@@ -143,51 +165,39 @@ export class Slider extends Focusable {
         return html`
             <div id="labelContainer">
-                <label id="label" for="input"><slot>${this.label}</slot></label>
+                <label id="label" for="input">${this.label}</label>
```
I'd love to find an acceptable way to allow use to keep this `<slot>`. In the combobox research, we've looked at having the `sp-combobox-item` elements sprout a `slot='option'` attribute to allow for this. Thoughts on `<sp-slider-handle>` doing similar? Maybe sprouting `slot="handle"`? This is also a way to support the `assignedElements` note above.

cuberoot replied:
Ok, giving that a try.

Westbrook commented on packages/slider/src/Slider.ts:
```diff
@@ -243,203 +253,60 @@ export class Slider extends Focusable {
+                    ${this.renderRamp()} ${this.renderTicks()}
+                    ${this.handleController.render()}
+                    ${segments
+                        .slice(1)
```
What are we slicing off here? Is there a way to do that in the controller?

cuberoot replied:
So, the CSS assumes that the first and last things rendered are track segments. So, I render the first segment explicitly and then render the remaining ones at the end. The `slice(1)` just takes all of the remaining segments after the first one.

Westbrook replied:
Interesting. I'd want to check whether slicing vs skipping had any effect on the way the template was addressed in subsequent renders, being the slice (IIRC) makes a new array identity every time.

cuberoot replied:
I reworked this to make sure that the various parts of rendered HTML have appropriate ids. I am not sure I gained much in performance. My ad hoc testing showed there was possibly a small difference.

Westbrook commented on packages/slider/src/SliderHandle.ts:
```diff
@@ -0,0 +1,69 @@
+    }
+}
+
+export class SliderHandle extends SliderHandleBase {
```
This likely wants a value for `focusElement`?

cuberoot replied:
Yes, I'll have to work that one out...

cuberoot replied:
Slider handles now supply a `focusElement` correctly

Westbrook commented on packages/slider/src/SliderHandle.ts:
```diff
@@ -0,0 +1,69 @@
+        return this.name;
+    }
+
+    public updated(changedProperties: PropertyValues): void {
```
Ah, this isn't in the controller, I see. Interesting.

Westbrook replied:
This never actually _renders_ anything, right? Might be a premature optimization, but in that case we might want to do this work in `shouldUpdate` and `return false;` to work within that context while reducing the call depth of an update.

cuberoot replied:
I'll change it.

Westbrook commented on packages/slider/stories/slider.stories.ts:
```diff
@@ -97,12 +143,122 @@ export const Disabled = (): TemplateResult => {
+                @change=${handleEvent}
+                ...=${spreadProps(args)}
+            >
+                <sp-slider-handle
+                    name="low"
+                    value="5"
+                    max="next"
+                ></sp-slider-handle>
+                <sp-slider-handle
+                    name="mid"
+                    value="100"
+                    min="previous"
+                    max="next"
+                ></sp-slider-handle>
+                <sp-slider-handle
+                    name="high"
+                    value="250"
+                    min="previous"
+                ></sp-slider-handle>
```
I'd love to see `handles` as a Storybook arg here, so that they could be dynamic to the story, but having all three versions available for the VRT is the most important part here, IMO.

Westbrook reviewed (commented):
@cuberoot this is such an exciting start! Most of this is more questions and encouragement than any really big requests, so hopefully it doesn't add too much to the process.

One "in the wild" usage to confirm here is in the VRT results project which is failing the builds: https://app.circleci.com/pipelines/github/adobe/spectrum-web-components/6745/workflows/bcc75a34-a311-460b-a448-3136b258304b/jobs/69421/parallel-runs/0/steps/0-103

Westbrook commented:
When you get to a place where it's possible, can you report the results of:
```
yarn build:tests
yarn test:bench -p slider
```
and
```
yarn test:bench -p slider -s page
```

cuberoot commented on packages/slider/sp-slider.ts:
```diff
@@ -10,11 +10,14 @@ OF ANY KIND, either express or implied. See the License for the specific languag
 declare global {
     interface HTMLElementTagNameMap {
         'sp-slider': Slider;
+        'sp-slider-handle': SliderHandle;
```
I'm happy to change if you think it is worth adding in an `sp-slider-handle.ts`.

Westbrook replied:
Will reserve judgement till the very end, but possibly.

Westbrook replied:
Speaking through this a little:

Having this together means that if you want to use an `sp-slider` element but for some reason want to use a custom/extended `sp-slider-handle` element (possibly one that closes over some of the complexities you've shown possible in the three handle demos), a consumer would _have_ to extend and use a new element name. This works fine because you leverage `this.host.querySelectorAll('[slot="handle"]')` in the model resolver, so I don't think we have a strong need to move this. It is a little outside of other usage, none of the other paired elements register in a single file, but I'm not sure it's that big of a deal.

If you were to read this and decide "let's move it to `sp-slider-handle.ts`!", do take into account that the new file would need to be added to the `exports` map in `package.json` were you to do that.

cuberoot replied:
Ok, broke down and created `sp-slider-handle.ts`

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,371 @@
+*/
+import { html, TemplateResult, ifDefined } from '@spectrum-web-components/base';
+import { streamingListener } from '@spectrum-web-components/base/src/streaming-listener.js';
+import { Slider } from './Slider';
+import { SliderHandle, HandleValues } from './SliderHandle';
```
No justification. I'm just not used to working on projects where the .js is left off. Easy fix.

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,371 @@
+        for (const handle of this.model.values()) {
+            const formatOptions =
+                handle.formatOptions ?? this.host.formatOptions;
+            const formattedValue = formatOptions
```
👍

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,371 @@
+    public get hasExplicitHandles(): boolean {
+        if (this.model.size === 0) return false;
+        if (this.model.size > 1) return true;
+        return this.model.values().next().value !== this.host;
```
If the user has explicitly defined handles, then that changes how the `value` attribute on the `sp-slider` behaves.

This comes back to an earlier suggestion of yours that I had implemented. I have a `value` property on the `sp-slider` and `sp-slider-handle` that was always a `number`. I had then added a second property called `values` on the `sp-slider` that had the structured data for all of the sliders. You asked if I could just change the shape of the `value` property.

So, at the moment, there is only a `value` property and it returns a number when there are no explicit `sp-slider-handles`. When there are, it returns the structured `HandleValues` result. This is what is causing the break in the wild that you mentioned above...

I wonder if it would be better to go back to having two attributes? What do you think?

Westbrook replied:
It makes perfect sense later on in the review when I actually see it used, so really just a comment it fine for me.

cuberoot replied:
Comment added

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,371 @@
+        return this.model.values().next().value !== this.host;
+    }
+
+    public get focusElement(): HTMLElement {
```
Yah, I wasn't sure what to do here. What I was going for was that, if you click on the label, you should be able to use the keyboard on the handle that was last moved (defaulting to the first handle). The `input` elements are always rendered in the same order, so you seem to be able to tab and shift-tab between them in the correct order. This is going to need some testing.

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,371 @@
+    }
+
+    private extractModelFromLightDom = () => {
+        let elements = [
+            ...this.host.querySelectorAll('sp-slider-handle'),
+        ] as SliderHandle[];
```
Definitely a more advanced use case than I had envisioned. Worth a follow up later once we get the simple case working.

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,371 @@
+
+        this.handleOrder = [];
+        for (const name of this.model.keys()) {
+            this.handleOrder.push(name);
```
Yep.

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,371 @@
+    };
+
+    public get activeHandle(): string {
+        return this.handleOrder[0];
```
This is dynamic. When a handle is activated, I remove it from `handleOrder` and `unshift` it to the front of the list.

Westbrook replied:
🤯 Great!

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,371 @@
+    }
+
+    private getHandleComponents(name: string): HandleComponents {
+        const handle = this.host.shadowRoot.querySelector(
+            `.handle[name="${name}"]`
+        ) as HTMLElement;
+        const input = handle.querySelector('input') as HTMLInputElement;
```
Yes, I have been missing having a way to get a ref to a rendered element. Good to hear that one is coming. I didn't want to complicate an already intense code change by adding caching here 😄

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,371 @@
+    }
+
+    public get count(): number {
+        return this.model.size;
```
Not sure what you are asking here. `model` is a `Map`, so `size` is going to be what I call. It is possible that `model` isn't the right name. I refactored this PR twice before submitting it. The `model` started out as JSON data and turned into a `Map` of the actual slider handles.

Westbrook replied:
Just looking at it is weird. `HandleController.count` reaches in to `HandleController.model.size`. Is `size` for both feels weird, I'm not worried about it that much.

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,371 @@
+         * Dispatch a synthetic pointerdown event to ensure that pointerdown
+         * handlers attached to the slider are invoked before input handlers
+         */
+        event.stopPropagation();
+        const syntheticPointerEvent = new PointerEvent('pointerdown', event);
+        handle.dispatchEvent(syntheticPointerEvent);
+
+        input.value = this.calculateHandlePosition(event).toString();
+        model.value = parseFloat(input.value);
+        this.host.requestUpdate();
+        this.dispatchInputEvent(model);
```
I have merged your changes across to `beginTrackDrag` in `HandleController`. I'll push that soon.

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,371 @@
+     */
+    private calculateHandlePosition(event: PointerEvent | MouseEvent): number {
+        if (!this.boundingClientRect) {
+            return 0; //return this.value;
```
I just moved to using a getter instead (DRY principle):
```typescript
   private get boundingClientRect(): DOMRect {
        return this.host.getBoundingClientRect();
    }
```

Westbrook replied:
One thing to take into account is that `.getBoundingClientRect()` requires style recalc, so being able to cache it is really valuable.

cuberoot replied:
Ok, added caching back in...

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,371 @@
+        }
+        return html`
+            <div
+                class=${classes.join(' ')}
```
Excellent, I thought there should be something like that, but I missed it in my original search

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,371 @@
+        const normalizedValues = values.map((value) =>
+            this.normalizeValue(value)
+        );
+        normalizedValues.unshift(0);
```
Lets say we have a slider with min=0 and max=10 and handles at 3 and 6. There will be three track segments: `[0,3], [3,6], and [6,10]`. Normalized (between 0 and 1) this will look like: `[0,0.3], [0.3,0.6], and [0.6,1]`.

The `normalizedValues.unshift(0)` puts the starting value of `0` into the list of segment start values. Then, we iterate over the start values of the segment and product tuples of the `[start, end]` values.

I will change a variable name and add a comment to make it more clear

cuberoot commented:
> When you get to a place where it's possible, can you report the results of:
> 
> ```
> yarn build:tests
> yarn test:bench -p slider
> ```

```
┌─────────────┬───────────────────┐
│   Benchmark │ slider:test-basic │
├─────────────┼───────────────────┤
│     Browser │ chrome-headless   │
│             │ 90.0.4430.93      │
├─────────────┼───────────────────┤
│ Sample size │ 50                │
└─────────────┴───────────────────┘

┌─────────┬────────────┬───────────────────┬───────────────────┬───────────────────┐
│ Version │ Bytes      │          Avg time │         vs remote │               vs  │
├─────────┼────────────┼───────────────────┼───────────────────┼───────────────────┤
│ remote  │ 352.12 KiB │ 69.16ms - 69.79ms │                   │            faster │
│         │            │                   │          -        │         25% - 26% │
│         │            │                   │                   │ 22.99ms - 23.83ms │
├─────────┼────────────┼───────────────────┼───────────────────┼───────────────────┤
│ <none>  │ 360.89 KiB │ 92.61ms - 93.16ms │            slower │                   │
│         │            │                   │         33% - 34% │          -        │
│         │            │                   │ 22.99ms - 23.83ms │                   │
└─────────┴────────────┴───────────────────┴───────────────────┴───────────────────┘


Update available!
The latest version of tachometer is 0.5.9
You are running version 0.5.8
See what's new at https://github.com/Polymer/tachometer/blob/master/CHANGELOG.md
slider:test-basic: 69.16ms - 69.79ms
slider:test-basic: 92.61ms - 93.16ms
✨  Done in 48.53s.
```

> 
> and
> 
> ```
> yarn test:bench -p slider -s page
> ```
```

┌─────────────┬───────────────────┐
│   Benchmark │ slider:test-basic │
├─────────────┼───────────────────┤
│     Browser │ chrome-headless   │
│             │ 90.0.4430.93      │
├─────────────┼───────────────────┤
│ Sample size │ 50                │
└─────────────┴───────────────────┘

┌─────────┬────────────┬─────────────────────┬───────────────────┬───────────────────┐
│ Version │ Bytes      │            Avg time │         vs remote │               vs  │
├─────────┼────────────┼─────────────────────┼───────────────────┼───────────────────┤
│ remote  │ 352.12 KiB │ 117.35ms - 118.22ms │                   │            slower │
│         │            │                     │          -        │         25% - 27% │
│         │            │                     │                   │ 23.68ms - 24.88ms │
├─────────┼────────────┼─────────────────────┼───────────────────┼───────────────────┤
│ <none>  │ 360.89 KiB │   93.09ms - 93.92ms │            faster │                   │
│         │            │                     │         20% - 21% │          -        │
│         │            │                     │ 23.68ms - 24.88ms │                   │
└─────────┴────────────┴─────────────────────┴───────────────────┴───────────────────┘


Update available!
The latest version of tachometer is 0.5.9
You are running version 0.5.8
See what's new at https://github.com/Polymer/tachometer/blob/master/CHANGELOG.md
slider:test-basic: 117.35ms - 118.22ms
slider:test-basic: 93.09ms - 93.92ms
✨  Done in 30.15s.
```

Westbrook commented:
@cuberoot these perf profiles are something I'd really like to see clarified before we could accept this into the main library. 

`yarn test:bench -p slider` analyzes upgrade and render time of the element, and here it's showing a 33% - 34% degradation of performance. I'd say a little slower (at the benefit of the added functionality) would be acceptable, but this is a pretty solid drop. Further, it seems exacerbated when you use `sp-slider-handle` to outline the one handle. In that case, I'm seeing the following (further degraded) perf locally:
```
┌─────────────┬───────────────────┐
│   Benchmark │ slider:test-basic │
├─────────────┼───────────────────┤
│     Browser │ chrome-headless   │
│             │ 90.0.4430.93      │
├─────────────┼───────────────────┤
│ Sample size │ 50                │
└─────────────┴───────────────────┘

┌─────────┬────────────┬─────────────────────┬───────────────────┬───────────────────┐
│ Version │ Bytes      │            Avg time │         vs remote │               vs  │
├─────────┼────────────┼─────────────────────┼───────────────────┼───────────────────┤
│ remote  │ 352.29 KiB │ 108.09ms - 115.22ms │                   │            faster │
│         │            │                     │          -        │         37% - 41% │
│         │            │                     │                   │ 66.46ms - 76.57ms │
├─────────┼────────────┼─────────────────────┼───────────────────┼───────────────────┤
│ <none>  │ 363.37 KiB │ 179.59ms - 186.75ms │            slower │                   │
│         │            │                     │         58% - 70% │          -        │
│         │            │                     │ 66.46ms - 76.57ms │                   │
└─────────┴────────────┴─────────────────────┴───────────────────┴───────────────────┘
```

`yarn test:bench -p slider -s page` outlines the full load, define, upgrade, render process, and I'm somewhat perplexed that it's showing about the same perf profile as that without load/define. I'm quite pleased to see this profile flip back and prefer the new version by 20% - 21%, however. It could be that this comes from splitting the JS into multiple imports, or from our `updateComplete` not being appropriately measured (which could effect both profiles). I'm not sure a concrete next step there just now, but something we'll need to revisit as the functionality here settles.

cuberoot commented:
Ok, this is going to have to sit for a bit. I have to take what I have so far and work with it for now. I'll try to come back to this after next week.

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,371 @@
+    }
+
+    private handleStyle(model: SliderHandle): string {
+        return `${this.host.isLTR ? 'left' : 'right'}: ${
```
Yep, `styleMap` came into play.

cuberoot commented on packages/slider/src/Slider.ts:
```diff
@@ -16,44 +16,69 @@ import {
-        return this._value;
+    constructor() {
+        super();
+        this.handleController = new HandleController(this);
```
No, that was a mental leftover of using an old proprietary framework where the semantics were different. There's no reason not to have the value in an initialiser above in modern JS

cuberoot commented:
@Westbrook I have made a lot of changes. My latest benchmark is below:

```
┌─────────────┬───────────────────┐
│   Benchmark │ slider:test-basic │
├─────────────┼───────────────────┤
│     Browser │ chrome-headless   │
│             │ 90.0.4430.212     │
├─────────────┼───────────────────┤
│ Sample size │ 50                │
└─────────────┴───────────────────┘

┌─────────┬────────────┬───────────────────┬──────────────────┬──────────────────┐
│ Version │ Bytes      │          Avg time │        vs remote │              vs  │
├─────────┼────────────┼───────────────────┼──────────────────┼──────────────────┤
│ remote  │ 353.82 KiB │ 76.91ms - 79.02ms │                  │           faster │
│         │            │                   │         -        │         9% - 12% │
│         │            │                   │                  │ 7.73ms - 10.65ms │
├─────────┼────────────┼───────────────────┼──────────────────┼──────────────────┤
│ <none>  │ 364.88 KiB │ 86.14ms - 88.16ms │           slower │                  │
│         │            │                   │        10% - 14% │         -        │
│         │            │                   │ 7.73ms - 10.65ms │                  │
└─────────┴────────────┴───────────────────┴──────────────────┴──────────────────┘
```

It is an improvement over the previous benchmark and I hope it is more acceptable. However, I did [cheat slightly](https://github.com/adobe/spectrum-web-components/blob/7901a0ce9d3754471b4acba11e43f37fd7c6aa77/packages/slider/test/benchmark/test-basic.ts#L25) to get there. It turns out that `Intl.formatNumber` is surprisingly expensive. It accounted for most of the degradation in performance. So, in order to get a better sense of the performance changes related to the structure of the component, I bypassed it in the performance test.

cuberoot commented:
@Westbrook I have updated the tests and docs and I think I have addressed all of your comments. Please take another close look at it as I have refactored heavily since your last review. It is a bit simpler now.

I did eventually decide to stop overloading the `value` attribute/property on `sp-slider`. I had it so that it had structured data when there were multiple handles and only a number when there was a single implicit handle. However, as you pointed out, the getter and setter for `value` causes issues (which I guess is why you removed them). Users will probably use the `value` attributes on the individual handles anyway. There's now a `values` property on the `sp-slider` that will return all of the values.

Westbrook commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,540 @@
+            [this.host.isLTR ? 'left' : 'right']: `${
+                model.normalizedValue * 100
+            }%`,
+            'z-index': zIndex.toString(),
```
Just an FYI, but, `styleMap()` does accept JS values for CSS properties, if you wanted.

cuberoot replied:
I'm not sure what change you are proposing here (although happy to learn!). 

```javascript
[this.host.isLTR ? 'left' : 'right']: `${
    model.normalizedValue * 100 
}%`,
```
This is a JS value, isn't it? I need the template string to stick the `%` on the end.

Westbrook replied:
Not that important, but you can do `zIndex: zIndex.toString(),`, IIRC. So, `zIndex` the JS property for the the property...never mind me.

Westbrook commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,540 @@
+    private updateModel(): void {
+        const handles = [...this.handles.values()];
+
+        const getRangeAndClamp = (index: number): RangeAndClamp => {
```
This is a pretty smart function. Almost enough to want to have it moved somewhere it could be individually unit tested. Not a blocker.

Westbrook commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,540 @@
+                <input
+                    type="range"
+                    class="input"
+                    id=${ifDefined(isActive ? 'input' : undefined)}
```
We likely should switch from the current `<label for="id"/><output for="id"/><input id="id" />` style of labeling to a `<label id="id1" /><label id="id1" /><input aria-labelledby="id1 id2" />` style so that any number of inputs can reference the `label` and `output` element at the same time.

cuberoot replied:
I am a bit confused. Your suggestion above seems to show multiple labels for one input. Whereas I we have multiple inputs for one label. Am I missing something? 😕

cuberoot replied:
Ok, that looks similar to what I did based on React Spectrum

Westbrook commented on packages/slider/src/Slider.ts:
```diff
@@ -228,194 +206,65 @@ export class Slider extends Focusable {
-                    ${this.renderTrackLeft()} ${this.renderRamp()}
-                    ${this.renderTicks()} ${this.renderHandle()}
-                    ${this.renderTrackRight()}
+                    ${repeat(
```
Is `repeat` to support them changing order? Or something else?

cuberoot replied:
Using `repeat` just allowed me to specifically assign an id to each element rendered to make sure that lit can track them. Previously I was using the `slice` directly in the rendering and you were concerned about that affecting the performance on subsequent renders.

Westbrook commented on packages/slider/src/Slider.ts:
```diff
@@ -124,52 +120,33 @@ export class Slider extends Focusable {
+                ${this.hideValueLabel
+                    ? ''
+                    : html`
+                          <output id="value" aria-live="off" for="input">
```
We should use something like the following on this instead here:
```
.visually-hidden {
  clip: rect(0 0 0 0);
  clip-path: inset(50%);
  height: 1px;
  overflow: hidden;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}
```
This will allow the output to still be available for the screen reader.

Westbrook replied:
I've often thought to add this to our `styles` package and then ended up not fully needing the rule in the project I was working on, for instance.

Westbrook commented on packages/slider/stories/slider.stories.ts:
```diff
@@ -97,12 +145,234 @@ export const Disabled = (): TemplateResult => {
+                @change=${handleEvent}
+                ...=${spreadProps(args)}
+            >
+                <sp-slider-handle slot="handle" value="5"></sp-slider-handle slot="handle">
```
Double `slot`.

Westbrook commented on packages/slider/stories/slider.stories.ts:
```diff
@@ -97,12 +145,234 @@ export const Disabled = (): TemplateResult => {
+                ...=${spreadProps(args)}
+            >
+                Output Levels
+                <sp-slider-handle slot="handle" name="min" value="5"></sp-slider-handle slot="handle">
```
`slot`

Westbrook commented on packages/slider/stories/slider.stories.ts:
```diff
@@ -97,12 +145,234 @@ export const Disabled = (): TemplateResult => {
+    tickStep: 10,
+};
+
+// This is a very complex example from an actual application.
```
Can you expand on how this work, I'll have no way of known if I break this in the future or not.

Also, what would you consider the "strangest" values this could produce? It might be good to set that as the default so that the VRT can support catching regressions here.

Westbrook replied:
It's quite cool that we can do this now, however!

cuberoot replied:
I am adding a long explanation. You can decide if it is worth keeping. I have changed the values so that they exercise the most complex part of the calculations by default.

Westbrook commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,540 @@
+            if (handle.max === 'next') {
+                if (next) {
+                    for (let j = index + 1; j < handles.length; j++) {
+                        const next = handles[j];
```
JS doesn't mind using `next` here in different scopes, but it's a bit hard to track why we have next but then get a new next 🤔

cuberoot replied:
Agreed. I don't like variable shadowing either. Good catch...fixed.

Westbrook commented on packages/slider/stories/slider.stories.ts:
```diff
@@ -97,12 +145,234 @@ export const Disabled = (): TemplateResult => {
+            result = ((1 - 2 * value) * (Math.sqrt(9.99) - 1) + 1) ** 2;
+        }
+        const formatOptions = {
+            maximumFractionDigits: 2,
```
You might want `minimumFractionDigits: 2` here as well so that the text doesn't jump around so much. Though I guess that makes it display `1.00` vs `1` which might also not be appropriate.

cuberoot replied:
Yep, that looks better...

Westbrook commented on packages/slider/test/benchmark/test-basic.ts:
```diff
@@ -22,5 +22,6 @@ measureFixtureCreation(html`
         max="100"
         label="Opacity"
         id="opacity-slider"
+        .getAriaHandleText=${(value: number) => value.toString()}
```
I appreciate leveraging this in a proof context, but we should gate as outlined above or test the raw perf here going forward.

cuberoot replied:
Removed

Westbrook commented on projects/vrt-compare/src/OnionSkinner.ts:
```diff
@@ -32,7 +32,7 @@ export class OnionSkinner extends SpectrumElement {
 
     private handleOnionInput(event: Event & { target: Slider }) {
-        this.onionLevel = event.target.value;
+        this.onionLevel = event.target.value as number;
```
🙏🏼

cuberoot replied:
We don't need this anymore since I changed `value` back to a simple number attribute.

Westbrook reviewed (commented):
This is very exciting, and looks even more awesome!

I need to get into the tests tomorrow. Seems like not a lot of test for the amount of changes, but I've not actually looked at it so far.

cuberoot commented on packages/slider/README.md:
````diff
@@ -81,6 +81,98 @@ import { Slider } from '@spectrum-web-components/slider';
+### Range
+
+```html
+<div style="width: 500px; margin: 12px 20px;">
````
To be honest, I'm not sure why that was there. Must have been a copy-paste mistake. Deleting that

cuberoot commented on packages/slider/src/SliderHandle.ts:
```diff
@@ -0,0 +1,133 @@
+        value: number,
+        formatOptions: Intl.NumberFormatOptions
+    ) => string = (value, formatOptions) => {
+        return new Intl.NumberFormat(navigator.language, formatOptions).format(
```
I implemented formatter caching

cuberoot commented on packages/slider/README.md:
````diff
@@ -81,6 +81,98 @@ import { Slider } from '@spectrum-web-components/slider';
 <sp-slider label="Slider Label - Disabled" variant="ramp" disabled></sp-slider>
 ```
 
+### Range
````
I added `slider-handle.md` as suggested. That's a good idea.

cuberoot commented on packages/slider/README.md:
```diff
@@ -81,6 +81,98 @@ import { Slider } from '@spectrum-web-components/slider';
+    >
+        Output Levels
+        <sp-slider-handle slot="handle" name="min" value="5"></sp-slider-handle slot="handle">
+        <sp-slider-handle slot="handle" name="max" value="250"></sp-slider-handle slot="handle">
```
Fixed.

cuberoot commented on packages/slider/README.md:
```diff
@@ -81,6 +81,98 @@ import { Slider } from '@spectrum-web-components/slider';
+        name="high"
+        value="250"
+        min="previous"
+    ></sp-slider-handle slot="handle">
```
Fixed

cuberoot commented on packages/slider/README.md:
```diff
@@ -81,6 +81,98 @@ import { Slider } from '@spectrum-web-components/slider';
+        value="5"
+        max="next"
+    ></sp-slider-handle slot="handle">
+    <sp-slider-handle slot="handle"
```
This was a regression. I have fixed it.

cuberoot commented:
I made some progress today. I'll tackle more tomorrow!

cuberoot commented on packages/slider/src/Slider.ts:
```diff
@@ -59,32 +58,52 @@ export class Slider extends Focusable {
 
-    @property({ attribute: false })
-    public getAriaValueText: (value: number) => string = (value) => `${value}`;
+    @property({ type: String, reflect: true })
+    public language: string = navigator.language;
```
As I mentioned in #1495, my application doesn't necessarily use the value from `navigator.language`. So, I added this here. It seems a little painful to have to provide this to each component though. I wonder if there's a better API we could provide to allow override of `navigator.language` for all SWC components on the page?

Westbrook replied:
Does this specifically need to reflect?

cuberoot replied:
Nope.

Westbrook commented on packages/slider/package.json:
```diff
@@ -24,7 +24,8 @@
         "./sp-slider": "./sp-slider.js",
-        "./sp-slider.js": "./sp-slider.js"
+        "./sp-slider.js": "./sp-slider.js",
+        "./sp-slider-handle.js": "./sp-slider-handle.js"
```
We should include the no extension on the left version as well.

cuberoot replied:
Good catch!

Westbrook commented on packages/slider/src/SliderHandle.ts:
```diff
@@ -0,0 +1,163 @@
+        ) {
+            this._numberFormatCache = {
+                language,
+                numberFormat: new Intl.NumberFormat(
```
We use https://www.npmjs.com/package/@internationalized/number to better align Number Field to other Adobe offerings, should we do the same here?

cuberoot replied:
Done.

cuberoot commented on packages/slider/README.md:
```diff
@@ -81,6 +81,98 @@ import { Slider } from '@spectrum-web-components/slider';
+
+Normalization is the process of converting a slider to a value between 0 and 1 where
+0 represents the minimum and 1 represents the maximum. See ["Three Handles Complex"
+example in the playground](storybook/index.html?path=/story/slider--three-handles-complex).
```
I am working around this by using an actual `<a target="_blank" ...>` tag to open it in another tab.

cuberoot commented on packages/slider/README.md:
````diff
@@ -81,6 +81,98 @@ import { Slider } from '@spectrum-web-components/slider';
+attribute.
+
+```html
+<sp-slider hide-value-label></sp-slider>
````
Adding label

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,540 @@
+        }
+
+        /* c8 ignore next */
+        return undefined;
```
Changing to an exception. We shouldn't hit here.

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,540 @@
+                    aria-disabled=${ifDefined(
+                        this.host.disabled ? 'true' : undefined
+                    )}
+                    aria-valuetext=${this.formattedValueForHandle(model)}
```
That is a good question. Are there some accessibility folks that we ask for advice? I think we need to give some indication of which handle you are interacting with...

Westbrook replied:
@jnurthen thoughts on where we might back up our labeling approach here?

jnurthen replied:
@Westbrook have you looked at the react-spectrum RangeSlider? https://react-spectrum.adobe.com/react-spectrum/RangeSlider.html

cuberoot replied:
I have pushed a reworking of the labels based on the React Spectrum implementation. A basic summary is:
- the `<input>` for each handle is labelled by the slider's label and by itself
- there is an `aria-label` attribute on the `sp-slider-handle` that is passed to the `input`

See the range slider example on the slider-handle docs page to play with it.

The one difference is that React Spectrum automatically adds the `aria-labels` "Minimum" and "Maximum" to the inputs on the range slider. I have not done that since our solution is a little more free-form. But I could add that default for the two handle case if it made sense.

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,540 @@
+                        this.host.disabled ? 'true' : undefined
+                    )}
+                    aria-valuetext=${this.formattedValueForHandle(model)}
+                    @change=${(event: Event) =>
+                        this.onInputChange(event, model)}
+                    @focus=${(event: Event) => this.onInputFocus(event, model)}
+                    @blur=${() => this.onInputBlur(model)}
```
I wouldn't have thought of that. Taking the model on the input and then referencing the class members directly does knock off a couple of ms.

cuberoot commented on packages/slider/src/HandleController.ts:
```diff
@@ -0,0 +1,540 @@
+                } else {
+                    /* c8 ignore next 3 */
+                    console.warn(
+                        'First slider handle cannot have attribute max="next"'
```
Ah yep!

Westbrook commented on packages/slider/src/SliderHandle.ts:
```diff
@@ -0,0 +1,170 @@
+    public formatOptions?: NumberFormatOptions;
+
+    @property({ type: String, attribute: 'aria-label' })
+    public ariaLabel?: string;
```
This need to surface as the `label` attribute as any `aria-*` attribute will feed directly into the accessibility tree. And, while we want that on the `<input>` element, we don't want that on the `sp-slider-handle` element.

Westbrook replied:
It would be great to use the attribute in the stories and documentation, too.

cuberoot replied:
Any suggestions here? We could have a `label` attribute on `sp-slider-handle` that is passed through to `aria-label` on the `input`. It's a bit confusing to the developer using the component.

I did have one example in the docs [here](https://github.com/adobe/spectrum-web-components/blob/cuberoot/multi-handle-slider/packages/slider/slider-handle.md#L35), but more should be added.

Westbrook replied:
`<my-el label="Text">` is generally what we do here across the library. 

Can you expand on the confusion? I'm not sure how it's a different amount of confusion that placing `aria-label` on an element and then it being applied to two different things in the accessibility tree.

![Screen reader reporting of the Maximum Handle in the demo](https://user-images.githubusercontent.com/1156657/120705603-072d4080-c486-11eb-9883-a31cf1ca9058.png)

Beyond the attribute naming, this seems to interact with the screen reader quite nicely! If we can at least have  both versions here: https://cuberoot-multi-handle-slider--spectrum-web-components.netlify.app/components/slider-handle leverage this attribute, I think we'll be good for iteration 1.

cuberoot replied:
Ok, I will make it so. The confusion I meant was setting a `label` attribute that would never be rendered (only for aria). However, I think that is the best solution

jnurthen reviewed (commented):
Has any thought been put into what the behaviour should be when the minimum becomes the maximum? Currently the user can switch the slider handles which creates a confusing situation. react-spectrum limits the minimum value to be less than the maximum value so this situation does not occur

cuberoot commented:
> Has any thought been put into what the behaviour should be when the minimum becomes the maximum? Currently the user can switch the slider handles which creates a confusing situation. react-spectrum limits the minimum value to be less than the maximum value so this situation does not occur

We allow the handles to slide past each other by default because it is actually a use case in our software. However, you can specify `min="previous"` or `max="next"` to have the handles be limited according to their neighbour.

Westbrook commented:
We're very close here, some small (hopefully 🤞🏼 ) final things:
- if I click and drag a handle the associated `<input>` is no longer focused, so a user can't immediate use the arrow keys to update the value after the initial drag, this is a regression. This is likely due to the way we've changed the ID relationship to the label. Previously we could click the label to achieve focus because it only related to a single input. Check out https://github.com/adobe/spectrum-web-components/pull/1140 for other things that could regress while addressing this.
- there seems to be a _slight_ variation on the vertical placement of the track when there are multiple handles:
![Misaligned track UI](https://user-images.githubusercontent.com/1156657/120876904-89e6f600-c581-11eb-807f-96452a1b7210.png)
- all of the storybook stories need to have labels (Explicit Handle, Three Handles Ordered, Three Handles Complex)

I promise that I'll stop looking after you get these last pieces into the PR 😅

cuberoot commented:
@Westbrook I have fixed these two issues:

* when you click and drag, the handle now stays focused and responds to key events. I made sure that my fix matches your original implementation, which calls `click()` on the label
* All storybook sliders now have labels

I could not reproduce the alignment issue. I made two identical sliders in storybook, except one had two handles. I flipped back-and-forth, but there was no difference. See below.

![ScreenFlow](https://user-images.githubusercontent.com/154097/121470637-9d2b1300-ca12-11eb-89c8-c7b322e19d22.gif)

Westbrook commented:
Exciting!

- label ✅ 
- arrow keys, seeing this work. there's a peculiarity where the `:focus-visible` styles don't trigger once you move to the arrow keys, which would be expected. I can accept that being something we look at in follow up work (if it's not something you can see a path towards addressing easily) as managing the heuristic there can be a bit of a pain. Please create an issue here if you decide not to address that just now.

I looked deeper on this layout question and you can see it explicitly between these two stories:
- https://cuberoot-multi-handle-slider--spectrum-web-components.netlify.app/storybook/index.html?path=/story/slider--two-handles
- https://cuberoot-multi-handle-slider--spectrum-web-components.netlify.app/storybook/index.html?path=/story/slider--three-handles-ordered

The track travels up 1px moving between the first and the second story, seemingly due to the `variant` attribute being moved to `range` between the two. This seems to be due to the way that `margin` is applied in this rule:

```
:host([dir="ltr"][variant="range"]) .track,
:host([dir="rtl"][variant="range"]) .track {
    padding-top: 0px;
    padding-bottom: 0px;
    padding-left: var(--spectrum-slider-track-middle-handleoffset);
    padding-right: var(--spectrum-slider-track-middle-handleoffset);
    margin: var(--spectrum-slider-range-track-reset);
}
```
Which applies `margin-top: 0;` and overrides the "standard" or expected styling in:
```
#fill, 
.track {
    height: var(--spectrum-slider-track-height);
    box-sizing: border-box;
    position: absolute;
    z-index: 1;
    top: calc(var(--spectrum-slider-height)/2);
    margin-top: calc(var(--spectrum-slider-track-height)/-2);
    pointer-events: none;
}
```
Which would compute to `margin-top: -1px;`. It's purely cosmetic, so I'd accept an issue to address this later as well.

Westbrook commented:
The above does ask the question, "Does `variant="range"` do anything for the element?". And, if the answer is "yes", should we resolve that `variant` value for the consumer when we recognize that there is more than one `sp-slider-handle` element present?

cuberoot commented:
Logged #1527 for track issue. I don't think I understand what you are saying about the arrow key issue to describe it in an issue. I may leave you to enter that one.

I have to put this down for other work now. If it needs more polish, then hopefully someone can push it over the top.

Westbrook commented:
@cuberoot thanks for all the hard work on this and the clean handoff. We'll take it from here and have a release for you, soon!
